### PR TITLE
Cleans metablock commit cycle

### PR DESCRIPTION
**Pull request by pgev**
Cleans interaction between `Consensus::precommitMetablock`, `Committee::registerCommitteeDecision` and `Consensus::commitMetablock`functions.

  - Removes `bool` return value from `Committee::enterCommittee`.
  - `Committee` registers its decision to `Consensus` once reaches
    quorum (only once).
  - Adds `onlyCommittee` modifier to `Consensus`
  - Adds `decisions` public storage variable to `Consensus` to track
    `Committees` decisions.
  - Fixes 'committee formation blocksegment' check requirement
    in `Consensus::formCommittee`.
  - Updates iterator variable type from `uint8` to `uint256` in
    `Consensus::formCommittee` (iterating through committee formation
    blocksegment) to prevent possible infinite loop.
  - Adds `Consensus::registerCommitteeDecision` function.
  - Renames `Consensus::commit` to `Consensus::commitMetablock`.
  - Removes `Core::assertPrecommit` function and replaces it by
    `Core::hashMetablock`function.
  - `Core::openMetablock` has less input parameters as most checks are
    moved to `Consensus::commitMetablock`. It does not assert input parameters
    and relies on `Consensus` to validate before calling.
  - Renames `Consensus::registerPrecommit`to `Consensus::precommitMetablock`.
  - Updates all unit tests according to implementation. Some unit tests are skipped
    for `Consensus::commitMetablock`.
  - Slight documentation improvements of `Consensus`.

### CODE REVIEW

File: contracts/consensus/Consensus.sol [hunk 1]
```diff
@@ -295,15 +305,15 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
...
+            block.number <= precommit.committeeFormationBlockHeight
+                .sub(COMMITTEE_FORMATION_LENGTH)
+                .add(256),
+            "Committee formation blocksegment is not in most recent 256 blocks."
         );
 
         uint256 segmentHeight = precommit.committeeFormationBlockHeight;
         bytes32[] memory seedGenerator = new bytes32[](uint256(COMMITTEE_FORMATION_LENGTH));
-        for (uint8 i = 0; i < COMMITTEE_FORMATION_LENGTH; i++) {
+        for (uint256 i = 0; i < COMMITTEE_FORMATION_LENGTH; i++) {
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]: Add using safe math for `i++`

File: contracts/consensus/Consensus.sol [hunk 2]
```diff
@@ -295,15 +305,15 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
         );
 
         require(
-            block.number
-                .sub(uint256(COMMITTEE_FORMATION_LENGTH))
-                .sub(uint256(256)) < precommit.committeeFormationBlockHeight,
-            "Committee formation blocksegment length must be in 256 most recent blocks."
+            block.number <= precommit.committeeFormationBlockHeight
+                .sub(COMMITTEE_FORMATION_LENGTH)
+                .add(256),
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]: `COMMITTEE_FORMATION_LENGTH` is uint8. We can change it to uint256.
Also, we should explicitly cast `256` to `uint256(256)`.

File: contracts/consensus/Consensus.sol [hunk 3]
```diff
@@ -697,6 +680,66 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
...
+
+        bytes32 decision = decisions[committee];
+
+        require(
+            _committeeLock == keccak256(abi.encode(decision)),
+            "Committee decision does not match with committee lock."
+        );
+
+        require(
+            decision == precommit,
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]: This comparison is incorrect, because `metablockHash == precommit`

File: contracts/core/Core.sol [hunk 4]
```diff
@@ -668,18 +644,6 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
     {
         assert(precommit != bytes32(0));
```
Comment (deepesh-kn) [inline contracts/core/Core.sol:626]: Maybe we should pass `_precommit` in the params and check if its equal to the `precommit`.

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
👍 Thanks for the changes.
There are few changes that will be handled in https://github.com/mosaicdao/mosaic-1/issues/92 so I have not added the comments related to that.
Only two notable change requests will be related to `decision == precommit` and using safe math, Otherwise looks good to me.

File: contracts/consensus/Consensus.sol (hunk 1 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: Done.

File: contracts/consensus/Consensus.sol (hunk 2 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: I've added explicit cast.
Why should we change COMMITTEE_FORMATION_LENGTH to uint256?

File: contracts/consensus/Consensus.sol (hunk 3 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: I am not sure what do you mean by incorrect here? Could you please elaborate more?

File: contracts/core/Core.sol (hunk 4 above)
Comment (pgev) [reply contracts/core/Core.sol:626]: Why do you think we should do it (it's not going to harm if we do, except paying a gas). The thinking here is that `Consensus` does all checks and as this function is guarded by `onlyConsensus` we do not need to do any check on that regard. What are your thoughts?

File: contracts/consensus/Consensus.sol (hunk 2 above)
Comment (deepesh-kn) [reply contracts/consensus/Consensus.sol]: ok, with `sub()` it gets cast to `uint256`.
In the following code
```
bytes32[] memory seedGenerator = new bytes32[](uint256(COMMITTEE_FORMATION_LENGTH));
        for (uint256 i = 0; i < COMMITTEE_FORMATION_LENGTH; i++) {
```
Here in for loop `COMMITTEE_FORMATION_LENGTH` is `uint8`, and `i`  is `uint256`
If everywhere we need to cast it to `uint256` then it is better to declare it as `uint256`?

File: contracts/consensus/Consensus.sol (hunk 3 above)
Comment (deepesh-kn) [reply contracts/consensus/Consensus.sol]: I mean to say that, `committeeLock` is sha3 of `decision`.
`_committeeLock == keccak256(abi.encode(decision)),`
And, `precommit == metablockHash`.

So this condition ` decision == precommit,` is incorrect.

File: contracts/core/Core.sol (hunk 4 above)
Comment (deepesh-kn) [reply contracts/core/Core.sol:626]: I think it is ok. We trust consensus. 👍

File: contracts/consensus/Consensus.sol (hunk 2 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: We do not actually need to cast here to uint256. Why do you think we should cast here?

File: contracts/consensus/Consensus.sol (hunk 3 above)
Comment (pgev) [reply contracts/consensus/Consensus.sol]: I see. Correct. Removed.

File: contracts/consensus/Consensus.sol (hunk 2 above)
Comment (deepesh-kn) [reply contracts/consensus/Consensus.sol]: It was just a suggestion. Here logically it is correct. So resolving it.

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**